case_studies/dc2_cataloging/utils/convnet.py

```diff
@@ -0,0 +1,162 @@
+        self.cv1 = ConvBlock(c1, ch, kernel_size=1, padding=0, use_group_norm=use_group_norm)
+        self.cv2 = ConvBlock(
+            ch, c2, kernel_size=3, padding=1, stride=1, use_group_norm=use_group_norm
+        )
+        self.add = shortcut and c1 == c2
+
+    def forward(self, x):
+        out = self.cv2(self.cv1(x))
+        return x + out if self.add else out
+
+
+class C3(nn.Module):
+    def __init__(self, c1, c2, n=1, shortcut=True, e=0.5, use_group_norm=False):
+        super().__init__()
+        ch = int(c2 * e)
+        self.cv1 = ConvBlock(c1, ch, kernel_size=1, padding=0, use_group_norm=use_group_norm)
+        self.cv2 = ConvBlock(c1, ch, kernel_size=1, padding=0, use_group_norm=use_group_norm)
+        self.cv3 = ConvBlock(2 * ch, c2, kernel_size=1, padding=0, use_group_norm=use_group_norm)
+        self.m = nn.Sequential(
+            *(Bottleneck(ch, ch, shortcut, e=1.0, use_group_norm=use_group_norm) for _ in range(n)),
+        )
+
+    def forward(self, x):
+        return self.cv3(torch.cat((self.m(self.cv1(x)), self.cv2(x)), 1))
+
+
+class FeaturesNet(nn.Module):
+    def __init__(self, n_bands, ch_per_band, num_features, double_downsample=True):
+        super().__init__()
+
+        nch_hidden = 64
+        nch_hidden_for_asinh_params = 64
+        self.preprocess3d = nn.Sequential(
+            nn.Conv3d(n_bands, nch_hidden, [ch_per_band, 5, 5], padding=[0, 2, 2]),
+            nn.BatchNorm3d(nch_hidden),
+            nn.SiLU(),
+        )
+        self.backbone = nn.Sequential(
+            ConvBlock(nch_hidden + nch_hidden_for_asinh_params, 64, kernel_size=5, padding=2),
+            nn.Sequential(*[ConvBlock(64, 64, kernel_size=5, padding=2) for _ in range(4)]),
+            ConvBlock(64, 128, stride=2),
+            nn.Sequential(*[ConvBlock(128, 128) for _ in range(5)]),
+            ConvBlock(128, num_features, stride=(2 if double_downsample else 1)),  # 4
+        )
+
+        self.asinh_preprocess = nn.Sequential(
+            nn.ZeroPad2d(padding=(0, 1, 1, 1)),
+            C3(2, nch_hidden_for_asinh_params, n=4, use_group_norm=True),
+            nn.Upsample(scale_factor=2, mode="nearest"),
+            ConvBlock(
+                nch_hidden_for_asinh_params, nch_hidden_for_asinh_params, use_group_norm=True
+            ),
+            nn.Upsample(scale_factor=5, mode="nearest"),
+            ConvBlock(
+                nch_hidden_for_asinh_params,
+                nch_hidden_for_asinh_params,
+                kernel_size=7,
+                padding=3,
+                use_group_norm=True,
+            ),
+        )
+
+    def forward(self, x):
+        preprocessed_x = self.preprocess3d(x[0]).squeeze(2)
+        asinh_params = x[1]
+
+        asinh_params_mean = asinh_params.mean(dim=2, keepdim=True)
+        asinh_params_var = asinh_params.var(dim=2, keepdim=True)
+        normalized_asinh_params = (asinh_params - asinh_params_mean) / torch.sqrt(
+            asinh_params_var + 1e-5
```

Comment:
avoid multiline statements when possible (typically by using more intermediate variables). this helps with readability.